File: mobile_flutter/lib/models/barber.dart
class Service {
  final String id;
  final String name;
  final double defaultPrice;

  Service({
    required this.id,
    required this.name,
    required this.defaultPrice,
  });

  factory Service.fromJson(Map<String, dynamic> json) {
    return Service(
      id: json['_id'] ?? '',
      name: json['name'] ?? '',
      defaultPrice: (json['defaultPrice'] ?? 0).toDouble(),
    );
  }
}

class Location {
  final String city;
  final String address;

  Location({required this.city, required this.address});

  factory Location.fromJson(Map<String, dynamic> json) {
    return Location(
      city: json['city'] ?? '',
      address: json['address'] ?? '',
    );
  }
}

class Barber {
  final String id;
  final String name;
  final String shopName;
  final Location location;
  final String? profilePhoto;
  final List<Service> services;

  Barber({
    required this.id,
    required this.name,
    required this.shopName,
    required this.location,
    this.profilePhoto,
    required this.services,
  });

  factory Barber.fromJson(Map<String, dynamic> json) {
    return Barber(
      id: json['_id'] ?? '',
      name: json['name'] ?? '',
      shopName: json['shopName'] ?? '',
      location: Location.fromJson(json['location'] ?? {}),
      profilePhoto: json['profilePhoto'],
      services: (json['services'] as List? ?? [])
          .map((s) => Service.fromJson(s))
          .toList(),
    );
  }
}

File: backend/src/models/Barber.js
import mongoose from 'mongoose';
import bcrypt from 'bcryptjs';

const barberSchema = new mongoose.Schema({
  name: { type: String, required: true },
  shopName: { type: String, required: true },
  email: { type: String, required: true, unique: true, lowercase: true },
  password: { type: String, required: true },
  mobile: { type: String, required: true },
  location: {
    city: { type: String, required: true },
    address: { type: String, required: true }
  },
  shopNumber: { type: String, required: true },
  businessLicense: { type: String, required: true },
  profilePhoto: { type: String }, // path to the uploaded image
  services: [{ type: mongoose.Schema.Types.ObjectId, ref: 'Service' }],
  fcmToken: { type: String } // Firebase Cloud Messaging Token
}, { timestamps: true });

// Hash password before saving
barberSchema.pre('save', async function (next) {
  if (!this.isModified('password')) {
    next();
  }
  const salt = await bcrypt.genSalt(10);
  this.password = await bcrypt.hash(this.password, salt);
});

// Method to compare password
barberSchema.methods.matchPassword = async function (enteredPassword) {
  return await bcrypt.compare(enteredPassword, this.password);
};

const Barber = mongoose.model('Barber', barberSchema);
export default Barber;
